### PR TITLE
compaction: change behaviour of compaction task executors

### PR DESCRIPTION
**Pull request by Deexie**
Compaction tasks executors serve two different purposes - as compaction
manager related entity they execute compaction operation and as task
manager related entity they track compaction status. 

When one role depends on the other, as it currently is for 
compaction_task_impl::done() and compaction_task_executor::compaction_done(),
requirements of both roles need to be satisfied at the same time in each 
corner case. Such complexity leads to bugs.

To prevent it, compaction_task_impl::done() of executors no longer depends 
on compaction_task_executor::compaction_done().

Fixes: #14912.

### CODE REVIEW

**Comment by tchaikov** [comment]
```
materialized_views_test.TestInterruptBuildProcess.test_interrupt_build_process_and_resharding_max_to_half_test
...
    def wait_for_view(cluster, session, ks, view, raise_exception=True, timeout=None, cl=None):
...
        if raise_exception:
>           raise Exception(error_msg)
E           Exception: View ks.t_by_v2 not built
```

seems it timed out when waiting for status update in the `view_build_status` table. not sure if it is related, retriggered the build just now.

**Comment by Deexie** [comment]
- rebase

**Review by bhalevy — APPROVED** [review]
LGTM.

@raphaelsc please review too

**Review by denesb — APPROVED** [review]

File: compaction/compaction_manager.cc [hunk 1]
```diff
@@ -548,12 +554,6 @@ requires (compaction_manager& cm, Args&&... args) {
...
     gate::holder gate_holder = task_executor->_compaction_state.gate.hold();
-    _tasks.push_back(task_executor);
-    auto unregister_task = defer([this, task_executor] {
-        _tasks.remove(task_executor);
-        task_executor->switch_state(compaction_task_executor::state::none);
-        task_executor->release_resources();
-    });
 
     if (parent_info) {
         auto task = co_await get_task_manager_module().make_task(task_executor, parent_info.value());
```
Comment (raphaelsc) [inline compaction/compaction_manager.cc:561]: if compaction manager is asked to remove a table (which means stopping all tasks associated with that table), wouldn't it miss a task that is currently sleeping in task manager's make_task (which can preempt)? Then it's undefined behavior what happens thereafter as table could be dropped by then.

You said that motivation for patch is "Registering them in compaction manager earlier can lead to undesired behaviour, e.g. appearing as they are complete". Cannot we special case tasks that are in pending state and show them as they haven't started yet?

Comment (Deexie) [reply compaction/compaction_manager.cc:561]: My motivation for the patch is quite vague. What I really meant was that there may be situation where `stop_tasks` won't wait for the compaction because compaction_done() will be resolved immediately even though it shouldn't.

Getting into details:
We have 3 main operations here:
1. calling make_task to register it in task manager (preempts)
2. insert task into compaction_manager::_tasks
3. compaction_task_executor::_compaction_done = do_run()

Let's try to order them.

First let's see that there can be no preemption betweem 2 and 3:
- 2 -> preemption (with compaction_manager::stop_ongoing_compactions call) -> 3 -
	2 
	compaction_task_executor::_compaction_done has a default value which is a ready future
	compaction_manager::stop_ongoing_compactions stops tasks and waits for them to be complete (`co_await task->compaction_done();` in `stop_tasks`)
	3 - and task runs even though it was supposed to stop
^ this is a situation which I was trying to avoid with the patch
- 3 -> preemption (with compaction_manager::stop_ongoing_compactions call) -> 2
	3
	compaction_manager::stop_ongoing_compactions does not know about the task so it's not stopped even though it is supposed to.
	2

So, let's consider them as a group [23] (order of the two does not matter) and try to find a good way for 1.

A) [23] -> 1
Compaction may be running while it isn't registered in task manager yet. Quite not hurtful. Quite not what we aim for.

1 -> [23]
Here's the problem you have mentioned. Actually I think it won't be UB but a really long table removal since in compaction_manager::remove we wait until 
`co_await seastar::when_all_succeed(stop_ongoing_compactions("table removal", &t), c_state.gate.close()).discard_result();` and the gate won't close until the unlucky compaction finishes and leaves the gate. Nevertheless, it's not acceptable.

To summarize: the only remaining option here is A), however it's rather not satisfying.

I think this leads to conclusion that composition, and not inheritance, of compaction_task_executor is the way and it shouldn't be postponed anymore. (@bhalevy what do you think?)

Comment (raphaelsc) [reply compaction/compaction_manager.cc:561]: I see. Maybe we can use compaction_task_executor::stopping() to decide if we want to proceed with compaction. Alternatively, we play with promise and future, where default becomes non ready future and resolved once task either fails or succeed, but that may add unneeded complexity. And that means relying on compaction starting to actually stop the task, and that's not needed.

Comment (Deexie) [reply compaction/compaction_manager.cc:561]: That's right, thanks, I don't know why abandoned other ways to fix 213.

**Comment by Deexie** [comment]
- rebase

**Comment by mykaul** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3563/

https://github.com/scylladb/scylladb/issues/15260
Unerlated (I think) - should re-trigger CI.

**Comment by denesb** [comment]
@raphaelsc was your concern addressed?

**Comment by raphaelsc** [comment]
> @raphaelsc was your concern addressed?

yes, lgtm
